### ISSUE TITLE
[TypeScript]: `min` and `max` are marked as required on IHeatRule, but are not necessary when providing a `customFunction`

### ISSUE THREAD

**Issue by lindapaiste** [amcharts/amcharts5#883]
**Bug description**

This is just a TypeScript issue, not a JS/runtime issue.

I'm trying to use a `customFunction` for setting a series `heatRules`, but I'm getting TypeScript errors if I don't provide a `min` and `max` -- even though these are not required and not used.  The [code in the docs](https://www.amcharts.com/docs/v5/concepts/settings/heat-rules/#custom-functions) has this problem as well.
```
series.set("heatRules", [{
  target: series.columns.template,
  dataField: "value",
  customFunction: function(sprite: am5.Sprite, min, max, value) {
    if (value < 100) {
      (sprite as am5.Graphics).set("fill", am5.color(0xff0000));
    }
    else {
      (sprite as am5.Graphics).set("fill", am5.color(0x00ff00));
    }
  }
}])
```
There will be a big red underline around the whole object with the error:
> TS2739: Type '{ target: Template ; dataField: string; customFunction: (sprite: Sprite, min: number, max: number, value: any) => void; }' is missing the following properties from type 'IHeatRule': min, max

**Solution?**
The simplest solution would be to mark both `min` and `max` as optional.  The downside to this would be that invalid arguments are now type-safe. A user could omit `min` and `max` without error even if there is no `customFunction` .

The other approach would be to make the types more descriptive.  The downside here is that it doesn't really fit the code style of this repo to do it "right".  It is possible to use a TypeScript union type to say that there must be *either* a `customFunction` *or* a `min` and `max`.  Something like this:
```
type IHeatRule = {
  /**
   * Target template.
   */
  target: Template<any>;
  /**
   * Which data field to use when determining item's value.
   */
  dataField: string;
  /**
   * Custom lowest value.
   */
  minValue?: number;
  /**
   * Custom highest value.
   */
  maxValue?: number;
} & ({
  /**
   * The setting value to use for items if the lowest value.
   */
  min: any;
  /**
   * The setting value to use for items if the highest value.
   */
  max: any;
  /**
   * The setting value to use for items which do not have value at all.
   */
  neutral?: any;	
  /**
   * A setting key to set.
   */
  key?: string;
  /**
   * Use logarithmic scale when calculating intermediate setting values.
   *
   * @default false
   */
  logarithmic?: boolean;
} | {
  /**
   * A custom function that will set target element's settings.
   *
   * Can be used to do custom manipulation on complex objects requiring more
   * than modifying a setting.
   */
  customFunction: (target: Sprite, minValue: number, maxValue: number, value?: any) => void;
});
```
It's broken into three sections like so:
```
type IHeatRule = CoreTypes & (MinMaxTypes | FunctionTypes);
```
My understanding is that the other properties in the `MinMaxTypes` section will not have any effect if a `customFunction` is provided...Though now I'm looking at the code and the [`if (rule.customFunction) {`](https://github.com/amcharts/amcharts5/blob/e134effe93e522ba15667ddd5c17662b5810a0b1/src/.internal/core/render/Series.ts#L608) check comes very late in the code, after a bunch of computations which would be unnecessary when `true`?  So I'm going to stop myself before falling any deeper into this rabbit hole.

Here's the outcome of the proposed complex types:
```
// Okay
series.set('heatRules', [{
  target: series.columns.template,
  dataField: 'value',
  customFunction: (sprite, min, max, value) => {
    // do something;
  }
}])

// Okay
series.set('heatRules', [{
  target: series.columns.template,
  dataField: 'value',
  min: 0,
  max: 100
}])

// Error
series.set('heatRules', [{
  target: series.columns.template,
  dataField: 'value'
}])
```
[TypeScript Playground Link](https://tsplay.dev/WyoQ6N)

**Comment by martynasma** [comment]
Those are good points. We'll consider the solutions. Thanks!

**Comment by martynasma** [comment]
Fixed in [5.3.12](https://github.com/amcharts/amcharts5/releases/tag/5.3.12).

## [5.3.12] - 2023-05-05

### Added
- New locale: Faroese (fo_FO). Thanks Martin Puge.

### Changed
- The `min` and `max` settings are now optional for `IHeatRule` ([Issue 883](https://github.com/amcharts/amcharts5/issues/883)).
- Removed `xlsx` package (with vulnerability) from dependencies in favor of a bundled hard copy (version 0.19.3) due to their decision not to publish to NPM. ([Issue 897](https://github.com/amcharts/amcharts5/issues/897)).

### Fixed
- Events now work properly with open shadow DOM (but not closed shadow DOM).
- Fixed export of data freezing in some cases where there are visual elements on the chart that are constantly being updated.
- Using parentheses in the in-line number formatting function was not working as needed.
- Tapping on a touch display outside chart was applying default state to elements that weren't being "touch-hovered".
- `getDataItemById()` method was not working properly on `Hierarchy` series.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
